File: src/testmd/report.py
from __future__ import annotations

import json
from datetime import datetime, timezone
from pathlib import Path

import click

from .models import TestInstance
from .resolver import StatusResult, changed_files

STATUS_ICONS = {
    "resolved": ("✓", "green"),
    "failed": ("✗", "red"),
    "outdated": ("⟳", "yellow"),
    "pending": ("…", "cyan"),
}


def print_status(results: list[StatusResult]) -> None:
    by_title: dict[str, list[StatusResult]] = {}
    for r in results:
        by_title.setdefault(r[0].definition.title, []).append(r)

    for title, group in by_title.items():
        click.echo(click.style(title, bold=True))
        for inst, status, rec in group:
            _print_instance_line(inst, status, rec)
        click.echo()

    _print_summary(results)


def print_get(inst: TestInstance, status: str, record: dict | None) -> None:
    title = _substitute_labels(inst.definition.title, inst.labels)
    click.echo(click.style(f"# {title}", bold=True))

    if inst.labels:
        click.echo(f"Labels: {format_labels(inst.labels)}")
    click.echo(f"Status: {_styled_status(status)}")

    if record:
        if record.get("resolved_at"):
            click.echo(f"Resolved at: {record['resolved_at']}")
        if record.get("failed_at"):
            click.echo(f"Failed at: {record['failed_at']}")
        if record.get("message"):
            click.echo(f"Message: {record['message']}")

    click.echo(f"Patterns: {', '.join(inst.resolved_patterns)}")
    click.echo(f"Files: {len(inst.matched_files)}")

    if status == "outdated":
        diff = changed_files(inst, record)
        if diff:
            click.echo(click.style("Changed:", fg="yellow"))
            for f in diff:
                click.echo(f"  {f}")

    click.echo("---")
    click.echo(_substitute_labels(inst.definition.description, inst.labels))


def write_report_md(results: list[StatusResult], path: Path) -> None:
    lines = ["# Test Report", ""]

    by_title: dict[str, list[StatusResult]] = {}
    for r in results:
        by_title.setdefault(r[0].definition.title, []).append(r)

    for title, group in by_title.items():
        lines.append(f"## {title}")
        lines.append("")
        lines.append("| ID | Labels | Status | Message |")
        lines.append("|----|--------|--------|---------|")
        for inst, status, rec in group:
            icon = STATUS_ICONS[status][0]
            labels = format_labels(inst.labels) or "—"
            msg = (rec or {}).get("message") or ""
            lines.append(f"| `{inst.id}` | {labels} | {icon} {status} | {msg} |")
        lines.append("")

    counts = _count(results)
    lines.append("## Summary")
    lines.append("")
    for s in ("resolved", "failed", "outdated", "pending"):
        lines.append(f"- {s}: {counts.get(s, 0)}")

    path.parent.mkdir(parents=True, exist_ok=True)
    path.write_text("\n".join(lines) + "\n")
    click.echo(f"Report saved to {path}")


def write_report_json(results: list[StatusResult], path: Path) -> None:
    data = {
        "tests": [
            {
                "id": inst.id,
                "title": inst.definition.title,
                "labels": inst.labels,
                "status": status,
                "message": (rec or {}).get("message"),
                "files": inst.matched_files,
            }
            for inst, status, rec in results
        ],
        "summary": _count(results),
    }
    path.parent.mkdir(parents=True, exist_ok=True)
    path.write_text(json.dumps(data, indent=2, ensure_ascii=False) + "\n")
    click.echo(f"Report saved to {path}")


def format_labels(labels: dict[str, str]) -> str:
    if not labels:
        return ""
    return " ".join(f"{k}={v}" for k, v in sorted(labels.items()))


def _substitute_labels(text: str, labels: dict[str, str]) -> str:
    for var, val in labels.items():
        text = text.replace(f"${var}", val)
    return text


def _print_instance_line(
    inst: TestInstance, status: str, rec: dict | None
) -> None:
    icon, color = STATUS_ICONS[status]
    parts = [
        "  ",
        click.style(icon, fg=color),
        " ",
        click.style(inst.id, dim=True),
    ]

    if inst.labels:
        parts.append("  " + format_labels(inst.labels))

    parts.append("  " + _styled_status(status))

    if status == "failed" and rec and rec.get("message"):
        parts.append(f'  "{rec["message"]}"')
    elif status in ("resolved", "failed") and rec:
        ts = rec.get("resolved_at") or rec.get("failed_at")
        ago = _time_ago(ts)
        if ago:
            parts.append(f"  ({ago})")

    click.echo("".join(parts))


def _styled_status(status: str) -> str:
    _, color = STATUS_ICONS[status]
    return click.style(status, fg=color)


def _time_ago(iso_str: str | None) -> str:
    if not iso_str:
        return ""
    dt = datetime.fromisoformat(iso_str)
    secs = int((datetime.now(timezone.utc) - dt).total_seconds())
    if secs < 0:
        return ""
    if secs < 60:
        return f"{secs}s ago"
    if secs < 3600:
        return f"{secs // 60}m ago"
    if secs < 86400:
        return f"{secs // 3600}h ago"
    return f"{secs // 86400}d ago"


def _print_summary(results: list[StatusResult]) -> None:
    counts = _count(results)
    parts = []
    for s, color in [("resolved", "green"), ("failed", "red"), ("outdated", "yellow"), ("pending", "cyan")]:
        n = counts.get(s, 0)
        parts.append(click.style(f"{n} {s}", fg=color))
    click.echo("Summary: " + ", ".join(parts))


def _count(results: list[StatusResult]) -> dict[str, int]:
    counts: dict[str, int] = {}
    for _, status, _ in results:
        counts[status] = counts.get(status, 0) + 1
    return counts

File: internal/report/report.go
package report

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"strings"
	"time"

	"github.com/fatih/color"

	"github.com/testmd/testmd/internal/models"
	"github.com/testmd/testmd/internal/resolver"
)

var statusStyles = map[string]struct {
	icon  string
	color *color.Color
}{
	"resolved": {"✓", color.New(color.FgGreen)},
	"failed":   {"✗", color.New(color.FgRed)},
	"outdated": {"⟳", color.New(color.FgYellow)},
	"pending":  {"…", color.New(color.FgCyan)},
}

// PrintStatus prints the status of all tests to stdout, grouped by source file.
func PrintStatus(results []models.StatusResult, root string) {
	groups := groupBySource(results, root)

	for _, sg := range groups {
		dim := color.New(color.Faint)
		dim.Println(sg.source)
		for _, g := range sg.defs {
			bold := color.New(color.Bold)
			bold.Printf("  %s\n", g.title)
			for _, r := range g.results {
				printInstanceLine(r)
			}
		}
		fmt.Println()
	}
	printSummary(results)
}

// PrintGet prints detailed info about a test instance.
func PrintGet(r models.StatusResult) {
	inst := r.Instance
	bold := color.New(color.Bold)

	title := substituteLabels(inst.Definition.Title, inst.Labels)
	bold.Printf("# %s\n", title)

	if len(inst.Labels) > 0 {
		fmt.Printf("Labels: %s\n", FormatLabels(inst.Labels))
	}
	s := statusStyles[r.Status]
	fmt.Printf("Status: %s\n", s.color.Sprint(r.Status))

	if r.Record != nil {
		if r.Record.ResolvedAt != nil {
			fmt.Printf("Resolved at: %s\n", *r.Record.ResolvedAt)
		}
		if r.Record.FailedAt != nil {
			fmt.Printf("Failed at: %s\n", *r.Record.FailedAt)
		}
		if r.Record.Message != nil {
			fmt.Printf("Message: %s\n", *r.Record.Message)
		}
	}

	fmt.Printf("Patterns: %s\n", strings.Join(inst.ResolvedPatterns, ", "))
	fmt.Printf("Files: %d\n", len(inst.MatchedFiles))

	if r.Status == "outdated" {
		diff := resolver.ChangedFiles(inst, r.Record)
		if len(diff) > 0 {
			color.Yellow("Changed:")
			for _, f := range diff {
				fmt.Printf("  %s\n", f)
			}
		}
	}

	fmt.Println("---")
	fmt.Println(substituteLabels(inst.Definition.Description, inst.Labels))
}

// FormatLabels formats labels as "key=val key2=val2" sorted by key.
func FormatLabels(labels map[string]string) string {
	if len(labels) == 0 {
		return ""
	}
	keys := make([]string, 0, len(labels))
	for k := range labels {
		keys = append(keys, k)
	}
	sort.Strings(keys)
	parts := make([]string, len(keys))
	for i, k := range keys {
		parts[i] = k + "=" + labels[k]
	}
	return strings.Join(parts, " ")
}

// WriteReportMD writes a markdown report to a file.
func WriteReportMD(results []models.StatusResult, path string) error {
	var sb strings.Builder
	sb.WriteString("# Test Report\n\n")

	groups := groupByDefinition(results)
	for _, g := range groups {
		sb.WriteString(fmt.Sprintf("## %s\n\n", g.title))
		sb.WriteString("| ID | Labels | Status | Message |\n")
		sb.WriteString("|----|--------|--------|--------|\n")
		for _, r := range g.results {
			s := statusStyles[r.Status]
			labels := FormatLabels(r.Instance.Labels)
			if labels == "" {
				labels = "—"
			}
			msg := ""
			if r.Record != nil && r.Record.Message != nil {
				msg = *r.Record.Message
			}
			sb.WriteString(fmt.Sprintf("| `%s` | %s | %s %s | %s |\n",
				r.Instance.ID, labels, s.icon, r.Status, msg))
		}
		sb.WriteString("\n")
	}

	sb.WriteString("## Summary\n\n")
	counts := countStatuses(results)
	for _, s := range []string{"resolved", "failed", "outdated", "pending"} {
		sb.WriteString(fmt.Sprintf("- %s: %d\n", s, counts[s]))
	}

	if err := os.MkdirAll(filepath.Dir(path), 0755); err != nil {
		return err
	}
	if err := os.WriteFile(path, []byte(sb.String()), 0644); err != nil {
		return err
	}
	fmt.Printf("Report saved to %s\n", path)
	return nil
}

// WriteReportJSON writes a JSON report to a file.
func WriteReportJSON(results []models.StatusResult, path string) error {
	type testEntry struct {
		ID      string            `json:"id"`
		Title   string            `json:"title"`
		Source  string            `json:"source"`
		Labels  map[string]string `json:"labels"`
		Status  string            `json:"status"`
		Message *string           `json:"message"`
		Files   []string          `json:"files"`
	}

	entries := make([]testEntry, len(results))
	for i, r := range results {
		var msg *string
		if r.Record != nil {
			msg = r.Record.Message
		}
		entries[i] = testEntry{
			ID:      r.Instance.ID,
			Title:   r.Instance.Definition.Title,
			Source:  r.Instance.Definition.SourceFile,
			Labels:  r.Instance.Labels,
			Status:  r.Status,
			Message: msg,
			Files:   r.Instance.MatchedFiles,
		}
	}

	data := map[string]interface{}{
		"tests":   entries,
		"summary": countStatuses(results),
	}

	body, err := json.MarshalIndent(data, "", "  ")
	if err != nil {
		return err
	}

	if err := os.MkdirAll(filepath.Dir(path), 0755); err != nil {
		return err
	}
	if err := os.WriteFile(path, append(body, '\n'), 0644); err != nil {
		return err
	}
	fmt.Printf("Report saved to %s\n", path)
	return nil
}

// --- grouping ---

type sourceGroup struct {
	source string
	defs   []defGroup
}

type defGroup struct {
	title   string
	defn    *models.TestDefinition
	results []models.StatusResult
}

func groupBySource(results []models.StatusResult, root string) []sourceGroup {
	type key struct {
		source string
		defn   *models.TestDefinition
	}
	sourceOrder := []string{}
	sourceMap := map[string][]defGroup{}

	for _, r := range results {
		defn := r.Instance.Definition
		src, _ := filepath.Rel(root, defn.SourceFile)
		src = filepath.ToSlash(src)

		defs, exists := sourceMap[src]
		if !exists {
			sourceOrder = append(sourceOrder, src)
		}

		if len(defs) > 0 && defs[len(defs)-1].defn == defn {
			defs[len(defs)-1].results = append(defs[len(defs)-1].results, r)
		} else {
			defs = append(defs, defGroup{
				title:   defn.Title,
				defn:    defn,
				results: []models.StatusResult{r},
			})
		}
		sourceMap[src] = defs
	}

	groups := make([]sourceGroup, len(sourceOrder))
	for i, src := range sourceOrder {
		groups[i] = sourceGroup{source: src, defs: sourceMap[src]}
	}
	return groups
}

type group struct {
	title   string
	defn    *models.TestDefinition
	results []models.StatusResult
}

func groupByDefinition(results []models.StatusResult) []group {
	var groups []group
	for _, r := range results {
		defn := r.Instance.Definition
		if len(groups) > 0 && groups[len(groups)-1].defn == defn {
			groups[len(groups)-1].results = append(groups[len(groups)-1].results, r)
		} else {
			groups = append(groups, group{
				title:   defn.Title,
				defn:    defn,
				results: []models.StatusResult{r},
			})
		}
	}
	return groups
}

func printInstanceLine(r models.StatusResult) {
	s := statusStyles[r.Status]
	dim := color.New(color.Faint)

	fmt.Print("    ")
	s.color.Print(s.icon)
	fmt.Print(" ")
	dim.Print(r.Instance.ID)

	if len(r.Instance.Labels) > 0 {
		fmt.Print("  " + FormatLabels(r.Instance.Labels))
	}

	fmt.Print("  ")
	s.color.Print(r.Status)

	if r.Status == "failed" && r.Record != nil && r.Record.Message != nil {
		fmt.Printf("  \"%s\"", *r.Record.Message)
	} else if (r.Status == "resolved" || r.Status == "failed") && r.Record != nil {
		ts := r.Record.ResolvedAt
		if ts == nil {
			ts = r.Record.FailedAt
		}
		if ts != nil {
			if ago := timeAgo(*ts); ago != "" {
				fmt.Printf("  (%s)", ago)
			}
		}
	}
	fmt.Println()
}

// PrintCI prints failing tests in the same grouped format as status.
func PrintCI(failing []models.StatusResult, root string) {
	groups := groupBySource(failing, root)
	for _, sg := range groups {
		dim := color.New(color.Faint)
		dim.Println(sg.source)
		for _, g := range sg.defs {
			bold := color.New(color.Bold)
			bold.Printf("  %s\n", g.title)
			for _, r := range g.results {
				printInstanceLine(r)
			}
		}
		fmt.Println()
	}
}

func printSummary(results []models.StatusResult) {
	counts := countStatuses(results)
	parts := []struct {
		status string
		c      *color.Color
	}{
		{"resolved", color.New(color.FgGreen)},
		{"failed", color.New(color.FgRed)},
		{"outdated", color.New(color.FgYellow)},
		{"pending", color.New(color.FgCyan)},
	}

	fmt.Print("Summary: ")
	for i, p := range parts {
		if i > 0 {
			fmt.Print(", ")
		}
		p.c.Printf("%d %s", counts[p.status], p.status)
	}
	fmt.Println()
}

func countStatuses(results []models.StatusResult) map[string]int {
	counts := map[string]int{}
	for _, r := range results {
		counts[r.Status]++
	}
	return counts
}

func substituteLabels(text string, labels map[string]string) string {
	for k, v := range labels {
		text = strings.ReplaceAll(text, "{"+k+"}", v)
	}
	return text
}

func timeAgo(isoStr string) string {
	t, err := time.Parse(time.RFC3339, isoStr)
	if err != nil {
		return ""
	}
	secs := int(time.Since(t).Seconds())
	if secs < 0 {
		return ""
	}
	if secs < 60 {
		return fmt.Sprintf("%ds ago", secs)
	}
	if secs < 3600 {
		return fmt.Sprintf("%dm ago", secs/60)
	}
	if secs < 86400 {
		return fmt.Sprintf("%dh ago", secs/3600)
	}
	return fmt.Sprintf("%dd ago", secs/86400)
}
